rust/stub/src/pe_loader.rs

```diff
@@ -0,0 +1,91 @@
+use goblin::pe::PE;
+use uefi::{
+    prelude::BootServices,
+    table::{
+        boot::{AllocateType, MemoryType},
+        Boot, SystemTable,
+    },
+    Handle, Status,
+};
+
+pub struct Image {
+    pub base: *const u8,
+    pub length: usize,
+    pub entry: extern "efiapi" fn(Handle, SystemTable<Boot>) -> Status,
+}
+
+#[cfg(target_arch = "x86_64")]
+fn flush_instruction_cache(_start: *const u8, _length: usize) {
+    // x86_64 mandates coherent instruction cache
+}
+
+/// Loads and relocates a trusted PE file, returning a structure containing its allocated virtual memory range and entry point.
+pub unsafe fn load(boot_services: &BootServices, file_data: &[u8]) -> uefi::Result<Image> {
+    let pe = PE::parse(file_data).expect("Malformed PE file");
+
+    // Allocate all memory the image will need in virtual memory.
+    // We follow shim here and allocate as EfiLoaderCode.
+    let mut pos = 0;
+    for section in &pe.sections {
+        assert!(section.virtual_address >= pos);
+        pos = section.virtual_address + section.virtual_size;
+    }
+    let base = boot_services.allocate_pages(
+        AllocateType::AnyPages,
+        MemoryType::LOADER_CODE,
+        (pos as usize + 0xfff) >> 12,
+    )? as *mut u8;
+
+    // Populate all sections in virtual memory.
+    let source = file_data.as_ptr();
+    for section in &pe.sections {
+        let copy_size = u32::min(section.virtual_size, section.size_of_raw_data) as usize;
+        unsafe {
+            core::ptr::copy_nonoverlapping(
+                source.add(section.pointer_to_raw_data as usize),
+                base.add(section.virtual_address as usize),
+                copy_size,
+            );
+            core::ptr::write_bytes(
+                base.add(section.virtual_address as usize + copy_size),
+                0,
+                section.virtual_size as usize - copy_size,
+            );
+        }
+    }
+
+    // Apply image base relocations.
```

Comment:
I'm not sure what's going on here. Does [relocations](https://docs.rs/goblin/0.6.0/goblin/pe/section_table/struct.SectionTable.html#method.relocations) not work?

Reply:
PE has two different kinds of relocations: object relocations are what you linked, and they only get used when linking object files. Image base relocations apply to an executable (or shared library) and use a different mechanism I parse here.

(Side note: using the thing you linked would "work" in the sense that the Linux kernel happens to load, since it does not contain meaningful image base relocations. So would not bothering at all about relocations. Things staying that way is not an assumption I'd be willing to make, since that would silently break things.)